[SPARK-8673] [launcher] API and infrastructure for communicating with child apps.

vanzin opened the pull request:
This change adds an API that encapsulates information about an app
launched using the library. It also creates a socket-based communication
layer for apps that are launched as child processes; the launching
application listens for connections from launched apps, and once
communication is established, the channel can be used to send updates
to the launching app, or to send commands to the child app.

The change also includes hooks for local, standalone/client and yarn
masters.

Commit:
[SPARK-8673] [launcher] API and infrastructure for communicating with child apps.

This change adds an API that encapsulates information about an app
launched using the library. It also creates a socket-based communication
layer for apps that are launched as child processes; the launching
application listens for connections from launched apps, and once
communication is established, the channel can be used to send updates
to the launching app, or to send commands to the child app.

Commit:
[SPARK-8673] [core] Add support for launcher lib handles.

This change adds a generic backend to communicate with launcher lib
handles, and hooks it up to the local and standalone client backends.

Commit:
[SPARK-8673] [yarn] Add support for launch lib handles.

This change uses the code added in the core/ module to connect
back to launcher lib handles, for both client and cluster modes
(basically the same code in the case of YARN).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

vanzin commented:
This is a WIP since I still want to have a second look at the API and implementation. Also, I separated things into 3 commits in case people think the change is too large to be reviewed in one shot.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35889/consoleFull) for   PR 7052 at commit [`bc7eed8`](https://github.com/apache/spark/commit/bc7eed807953cfeaab5b6ab68e80f6673faeb96f).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #35889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35889/console) for   PR 7052 at commit [`bc7eed8`](https://github.com/apache/spark/commit/bc7eed807953cfeaab5b6ab68e80f6673faeb96f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

vanzin commented:
> This patch adds the following public classes (experimental):

Most of those are package-private, seems like the script needs some tweaking.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/local/LocalBackend.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37152 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37152/consoleFull) for   PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b).

AmplabJenkins commented:
Merged build triggered.

SparkQA commented:
**[Test build #37152 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37152/console)**     for PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
Jenkins retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37237 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37237/consoleFull) for   PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #16 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/16/consoleFull) for   PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b).

SparkQA commented:
[Test build #16 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/16/console) for   PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37237 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37237/console) for   PR 7052 at commit [`9798293`](https://github.com/apache/spark/commit/9798293c244ac8518222c0c9c2cb2053a4b6bd7b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/local/LocalBackend.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37259/consoleFull) for   PR 7052 at commit [`27079ec`](https://github.com/apache/spark/commit/27079ece2ef1421ced3161bd68af764202ca83c2).

Commit:
Minor API and javadoc changes.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37277/consoleFull) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).

SparkQA commented:
[Test build #37259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37259/console) for   PR 7052 at commit [`27079ec`](https://github.com/apache/spark/commit/27079ece2ef1421ced3161bd68af764202ca83c2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #37277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37277/console) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
Jenkins retest this please.

vanzin commented:
The last failures seem unrelated, so I'll remove the "WIP" since I think I'm content with the current state of the patch.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37297/consoleFull) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #19 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/19/consoleFull) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).

SparkQA commented:
[Test build #19 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/19/console) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37297/console) for   PR 7052 at commit [`33fdf9a`](https://github.com/apache/spark/commit/33fdf9a058ba9a89ffb7e05ec0a244a7bdfeb011).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38740/consoleFull) for   PR 7052 at commit [`267501a`](https://github.com/apache/spark/commit/267501a82128520bb4123c4b1dbc20c5c25fb59b).

SparkQA commented:
[Test build #38740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38740/console) for   PR 7052 at commit [`267501a`](https://github.com/apache/spark/commit/267501a82128520bb4123c4b1dbc20c5c25fb59b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Update local-cluster memory setting.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38771 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38771/consoleFull) for   PR 7052 at commit [`17697e5`](https://github.com/apache/spark/commit/17697e50c3abdea922b21f1cbb42da0c3a07076c).

SparkQA commented:
[Test build #38771 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38771/console) for   PR 7052 at commit [`17697e5`](https://github.com/apache/spark/commit/17697e50c3abdea922b21f1cbb42da0c3a07076c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java
	launcher/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41383/consoleFull) for   PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf).

SparkQA commented:
**[Test build #41383 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41383/console)**     for PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41383/
Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41459/consoleFull) for   PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf).

SparkQA commented:
[Test build #41459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41459/console) for   PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41459/
Test FAILed.

vanzin commented:
weird pyspark errors, not sure how they could be related. so, retest this please.

vanzin commented:
I think now that 1.5 is winding down, it's time to start pinging people directly... @andrewor14 @squito any chance you guys could take a look?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41472/consoleFull) for   PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf).

squito commented on core/src/test/resources/log4j.properties:
```diff
@@ -16,13 +16,19 @@
+log4j.appender.console.target=System.err
+log4j.appender.console.layout=org.apache.log4j.PatternLayout
+log4j.appender.console.layout.ConversionPattern=%t: %m%n
+
```
mean to be merged?  If we really do want this in here, should probably add a brief comment that developers can use this by configuring `test.appender`

squito commented on core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala:
```diff
@@ -0,0 +1,81 @@
+      .setSparkHome(sys.props("spark.test.home"))
+      .setConf(SparkLauncher.DRIVER_EXTRA_CLASSPATH, System.getProperty("java.class.path"))
+      .setConf("spark.ui.enabled", "false")
+      .setConf(SparkLauncher.DRIVER_EXTRA_JAVA_OPTIONS, s"-Dtest.appender=console")
```
oh I see now ... well I guess then the log4j file should have a brief comment saying its used for this test.  I guess this is just to keep them separated?

squito commented on launcher/src/main/java/org/apache/spark/launcher/ChildProcAppHandle.java:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  void setState(State s) {
+    if (!state.isFinal()) {
```
rather than being a no-op if the state is already final, shoudl it throw an exception?  Are there OK cases for trying to change the case after its final?

vanzin replied:
Throwing an exception here wouldn't help much; this method is called by the socket-handling thread, not a user thread, so the side-effect would be a log message and the connection to the app being closed.

I can add the log message, but I'd rather avoid closing the socket, in case the backend has a legitimate reason for sending the message.

squito commented on launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java:
```diff
@@ -323,6 +341,40 @@ public SparkLauncher setVerbose(boolean verbose) {
+   *
+   * @since 1.5.0
+   * @param listeners Listeners to add to the handle before the app is launched.
+   * @return A handle for the launched application.
```
can you add a comment on `launch()` that `startApplication()` is generally preferred?

squito commented on launcher/src/main/java/org/apache/spark/launcher/package-info.java:
```diff
@@ -45,5 +45,33 @@
  * </pre>
+ *
+ * <p>
+ * Since Spark 1.5.0, the {@link org.apache.spark.launcher.SparkLauncher#startApplication(
```
I think this should go above the section on `launch`, the `launch` section should clearly state its the old way of doing things.

squito commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/BaseYarnClusterSuite.scala:
```diff
@@ -144,7 +146,12 @@ abstract class BaseYarnClusterSuite
+
+    val propsFile = createConfFile(childClasspath = childClasspath, extraConf = submitConf)
     val writer = new OutputStreamWriter(new FileOutputStream(propsFile), UTF_8)
     props.store(writer, "Spark properties.")
```
I'm confused -- those props related to SPARK-4267 are already set above on line 132, and then they are written to the file here.  It looks like you are ignoring `props` when initially writing the file w/ `createConfFile`, but then overwriting it again with this call to `props.store`.

vanzin replied:
Probably leftover of the merge. I'll clean this up.

squito commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:
```diff
@@ -101,6 +106,37 @@ class YarnClusterSuite extends BaseYarnClusterSuite {
+      .startApplication()
+
+    try {
+      eventually(timeout(30 seconds), interval(100 millis)) {
```
these timeouts are really long ... does it really take so long?  I guess a long timeout isn't so bad if it regularly completes quickly to help prevent flakiness

vanzin replied:
It doesn't take nearly that long, but I don't trust shared build boxes.

SparkQA commented:
**[Test build #41472 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41472/console)**     for PR 7052 at commit [`3c78f1c`](https://github.com/apache/spark/commit/3c78f1c657e8a7747196e7750e310cd6c502cccf)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41472/
Test FAILed.

squito commented on core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala:
```diff
@@ -0,0 +1,81 @@
+
+    try {
+      eventually(timeout(10 seconds), interval(100 millis)) {
+        handle.getAppId() should not be (null)
```
seems like this will be a pretty common use -- maybe we should give the user a `waitForAppId(timeout)` convenience method.

vanzin replied:
Not sure. That's why I added the listener API, so that users can figure out the best way for them to monitor the application. Adding an explicit blocking API for that covers just a single use case, and one that I'd expect to actually not be that common.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

Commit:
Feedback.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41553/consoleFull) for   PR 7052 at commit [`8d3111b`](https://github.com/apache/spark/commit/8d3111bf0ebf27f824ce3c51efc926170993fa51).

SparkQA commented:
[Test build #41553 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41553/console) for   PR 7052 at commit [`8d3111b`](https://github.com/apache/spark/commit/8d3111bf0ebf27f824ce3c51efc926170993fa51).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41553/
Test PASSed.

Commit:
Add some explanation about how the server works.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41656 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41656/consoleFull) for   PR 7052 at commit [`8608f56`](https://github.com/apache/spark/commit/8608f561be5a2ea3bc8981e9756ae127f4bcdd98).

SparkQA commented:
[Test build #41656 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41656/console) for   PR 7052 at commit [`8608f56`](https://github.com/apache/spark/commit/8608f561be5a2ea3bc8981e9756ae127f4bcdd98).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41656/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	yarn/src/test/scala/org/apache/spark/deploy/yarn/BaseYarnClusterSuite.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

Commit:
Remove dependency on new class added for launcher tests (merge issue).

Commit:
Various things.

- Cleanup more code after last merge.
- Add notification in LauncherBackend for when the launcher disconnets,
  track whether launcher is connected.
- Fix yarn-client; make YarnClusterSuite launcher test use client
  mode since it covers more ground in this case.
- Make LauncherBackend a class since Scala gets confused when traits
  have private vars / vals.

Commit:
Merge branch 'master' into SPARK-8673

Commit:
Report a final state in yarn-client mode.

Commit:
Add javadoc, timeout test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42143/consoleFull) for   PR 7052 at commit [`e697824`](https://github.com/apache/spark/commit/e6978243e7e832c95c3e8566436b1cbc2771f730).

Commit:
Use the launcher library to drive YarnClusterSuite and friends.

This increases coverage of this code by running all integration tests
through the library, instead of using spark-submit directly.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42148/consoleFull) for   PR 7052 at commit [`6cfef3a`](https://github.com/apache/spark/commit/6cfef3afe8c037071431d190e70bc2cb762472d1).

SparkQA commented:
[Test build #42148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42148/console) for   PR 7052 at commit [`6cfef3a`](https://github.com/apache/spark/commit/6cfef3afe8c037071431d190e70bc2cb762472d1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42148/
Test FAILed.

SparkQA commented:
[Test build #42143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42143/console) for   PR 7052 at commit [`e697824`](https://github.com/apache/spark/commit/e6978243e7e832c95c3e8566436b1cbc2771f730).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42143/
Test FAILed.

Commit:
More docs + a style fix.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42153/consoleFull) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).

SparkQA commented:
[Test build #42153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42153/console) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42153/
Test FAILed.

vanzin commented:
streaming failure. retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42163/consoleFull) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).

SparkQA commented:
[Test build #42163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42163/console) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42163/
Test FAILed.

vanzin commented:
again unrelated failures (sadface). retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42173 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42173/consoleFull) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).

SparkQA commented:
[Test build #42173 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42173/console) for   PR 7052 at commit [`3ac46bf`](https://github.com/apache/spark/commit/3ac46bfd47f951babb2974fbf4762b0922484b21).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42173/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42494/consoleFull) for   PR 7052 at commit [`bf92906`](https://github.com/apache/spark/commit/bf929069b495725e5d0bd85bad447f99019d1af7).

SparkQA commented:
[Test build #42494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42494/console) for   PR 7052 at commit [`bf92906`](https://github.com/apache/spark/commit/bf929069b495725e5d0bd85bad447f99019d1af7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42494/
Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42503/consoleFull) for   PR 7052 at commit [`bf92906`](https://github.com/apache/spark/commit/bf929069b495725e5d0bd85bad447f99019d1af7).

SparkQA commented:
[Test build #42503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42503/console) for   PR 7052 at commit [`bf92906`](https://github.com/apache/spark/commit/bf929069b495725e5d0bd85bad447f99019d1af7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42503/
Test PASSed.

vanzin commented:
Alright, I think this is in pretty good shape now. Pinging @squito and @andrewor14 again.

squito commented on launcher/src/main/java/org/apache/spark/launcher/package-info.java:
```diff
@@ -17,17 +17,45 @@
+ *         .setMaster("local")
+ *         .setConf(SparkLauncher.DRIVER_MEMORY, "2g")
+ *         .startApplication();
+ *       handle.stop();
```
I wouldn't include a call to `handle.stop()` in this example -- or if you do, add a comment explaining why you would call it.

squito commented on launcher/src/main/java/org/apache/spark/launcher/OutputRedirector.java:
```diff
@@ -0,0 +1,74 @@
+        redirect();
+      }
+    });
+    this.sink = Logger.getLogger(getClass().getName());
```
maybe it would be useful to the user to control the name of this logger?  especially if they launch multiple spark apps, might want the logs to be distinct.

Also it would be good to document somewhere how the child process handles both stdout & also where its logs go.

squito commented:
lgtm, just some very minor comments

Commit:
Merge branch 'master' into SPARK-8673

Commit:
Feedback. Allow logger name for child proc output to be configured.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43041/consoleFull) for   PR 7052 at commit [`c0a1766`](https://github.com/apache/spark/commit/c0a176636b00c43bc571bbe5539772200cb48278).

SparkQA commented:
[Test build #43041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43041/console) for   PR 7052 at commit [`c0a1766`](https://github.com/apache/spark/commit/c0a176636b00c43bc571bbe5539772200cb48278).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43041/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43065/consoleFull) for   PR 7052 at commit [`80599a4`](https://github.com/apache/spark/commit/80599a49a28e48aad2cd2b034f903ea14d922a6f).

SparkQA commented:
[Test build #43065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43065/console) for   PR 7052 at commit [`80599a4`](https://github.com/apache/spark/commit/80599a49a28e48aad2cd2b034f903ea14d922a6f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43065/
Test FAILed.

Commit:
Borked merge.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43066/consoleFull) for   PR 7052 at commit [`a7e4d17`](https://github.com/apache/spark/commit/a7e4d17cf34ea5977f27456d0b8d7fc954463823).

SparkQA commented:
[Test build #43066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43066/console) for   PR 7052 at commit [`a7e4d17`](https://github.com/apache/spark/commit/a7e4d17cf34ea5977f27456d0b8d7fc954463823).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43066/
Test PASSed.

Commit:
Better logger prefix.

AmplabJenkins commented:
Merged build triggered.

SparkQA commented:
[Test build #43093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43093/consoleFull) for   PR 7052 at commit [`a582e12`](https://github.com/apache/spark/commit/a582e12c50383070c474cbab102699679a62e757).

SparkQA commented:
[Test build #43093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43093/console) for   PR 7052 at commit [`a582e12`](https://github.com/apache/spark/commit/a582e12c50383070c474cbab102699679a62e757).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43093/
Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43098/consoleFull) for   PR 7052 at commit [`a582e12`](https://github.com/apache/spark/commit/a582e12c50383070c474cbab102699679a62e757).

SparkQA commented:
[Test build #43098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43098/console) for   PR 7052 at commit [`a582e12`](https://github.com/apache/spark/commit/a582e12c50383070c474cbab102699679a62e757).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43098/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-8673

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43298/consoleFull) for   PR 7052 at commit [`7c53f3d`](https://github.com/apache/spark/commit/7c53f3db4adf793c1f3cb7ed762b0686134af49e).

SparkQA commented:
[Test build #43298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43298/console) for   PR 7052 at commit [`7c53f3d`](https://github.com/apache/spark/commit/7c53f3db4adf793c1f3cb7ed762b0686134af49e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43298/
Test FAILed.

vanzin commented:
Test failures seem to be caused by SPARK-10964. I'll rerun the tests once that one is fixed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -36,6 +37,9 @@ private[spark] class SparkDeploySchedulerBackend(
 
   private var client: AppClient = null
   private var stopping = false
+  private var launcherBackend = new LauncherBackend() {
```
change to a `val`

Commit:
var -> val

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43335 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43335/consoleFull) for   PR 7052 at commit [`97e5b38`](https://github.com/apache/spark/commit/97e5b386d8a591b3fb48fb3da6e94a4b3725f792).

SparkQA commented:
[Test build #43335 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43335/console) for   PR 7052 at commit [`97e5b38`](https://github.com/apache/spark/commit/97e5b386d8a591b3fb48fb3da6e94a4b3725f792).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43335/
Test FAILed.

vanzin commented:
The tests fail because of the netty RPC backend. They pass locally when I revert the default to akka. I'll investigate that before pushing this.

vanzin commented:
SPARK-10987 is the new bug to watch...

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43404/consoleFull) for   PR 7052 at commit [`97e5b38`](https://github.com/apache/spark/commit/97e5b386d8a591b3fb48fb3da6e94a4b3725f792).

SparkQA commented:
[Test build #43404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43404/console) for   PR 7052 at commit [`97e5b38`](https://github.com/apache/spark/commit/97e5b386d8a591b3fb48fb3da6e94a4b3725f792).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ChildProcAppHandle implements SparkAppHandle `
  * `abstract class LauncherConnection implements Closeable, Runnable `
  * `final class LauncherProtocol `
  * `  static class Message implements Serializable `
  * `  static class Hello extends Message `
  * `  static class SetAppId extends Message `
  * `  static class SetState extends Message `
  * `  static class Stop extends Message `
  * `class LauncherServer implements Closeable `
  * `class NamedThreadFactory implements ThreadFactory `
  * `class OutputRedirector `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43404/
Test PASSed.

squito commented:
lgtm,  merging to master!
